### PR TITLE
Custom resolver

### PR DESCRIPTION
**Pull request by rbuckton**
Adds support for passing a custom `resolveFile` to any `Host`.

/cc @bakkot

### CODE REVIEW

**Comment by bakkot** [comment]
(See https://github.com/rbuckton/grammarkdown/issues/46#issuecomment-612733375)

This is neat, but I don't have enough confidence I will be able to use this correctly - if I fail to pass a method somewhere, now I'm hitting disk or network. I'm not comfortable using anything which falls back to doing actual IO instead of throwing. (Also it would still end up referencing Node's builtins, which I also don't want.)

**Comment by rbuckton** [comment]
Just loading the package currently at least _imports_ `path`, even if `path.resolve` is deferred. I could also add a `SyntheticSyncHost` that essentially just mandates that those exist. Its similar to the `SingleFileSyncHost` which accepts an *optional* fallback host and throws for any IO-like operation if it wasn't provided.

**Comment by bakkot** [comment]
> Just loading the package currently at least imports path, even if `path.resolve` is deferred.

Most bundlers will eliminate the import, as long as you don't actually use any class or function which references it (even in a dead path within that class or function).

>  I could also add a `SyntheticSyncHost` that essentially just mandates that those exist.

That would work as long as none of its superclasses referenced any of Node's built-ins. It seems a little odd from a design point of view, though: reading from a string or an array is no more synthetic than reading from the disk or network.

It seems to me that the base `Host` type doesn't need to have any responsibilities except to declare what the Host interface is, and possibly to provide some helper methods implemented in terms of that interface. It is weird to me that, in addition, the base `Host` type is entangled with Node, which is a very specific kind of `Host` - it couples the interface with a particular implementation.

I would expect the base `Host` type to _require_ passing in read/resolve/write methods (or, since this is typescript, declare those to be abstract methods), and then `SyncNodeHost` would pass Node's sync read/resolve/write methods, `StringHost` would pass `file => { if (file !== this.file) throw 'bad'; return this.content; }` etc, and custom hosts could do the same with their read/resolve/write methods. That seems like it draws the right boundaries.

**Comment by rbuckton** [comment]
I agree with most of your suggestions around the various `Host` abstractions. Changing the names would unfortunately be a breaking API change, which I'd like to avoid.

I should have an update with most of these changes up shortly.

**Comment by rbuckton** [comment]
I've refactored the `Host` API a bit:

- `HostBase` - Common functionality between sync and async hosts. No NodeJS-specific dependencies.
- `CoreSyncHost` - Common API for a synchronous host. Supports user-defined behavior and host fallback. No NodeJS-specific dependencies. Derived from `HostBase`.
- `CoreAsyncHost` - Common API for an asynchronous host. Supports user-defined behavior and host fallback. No NodeJS-specific dependencies. Derived from `HostBase`.
- `StringSyncHost` - An implementation of `CoreSyncHost` for reading a single file. No NodeJS-specific dependencies. Derived from `CoreSyncHost`.
- `StringAsyncHost` - An implementation of `CoreAsyncHost` for reading a single file. No NodeJS-specific dependencies. Derived from `CoreAsyncHost`.
- `NodeSyncHost` - An implementation of `CoreSyncHost` for reading from the file system using No NodeJS APIs. Derived from `CoreSyncHost`. Alias: `SyncHost`.
- `NodeAsyncHost` - An implementation of `CoreAsyncHost` for reading from the file system using No NodeJS APIs. Derived from `CoreAsyncHost`. Alias: `AsyncHost`.

For a user-defined synchronous host, you can simply do:

```ts
const host = new CoreSyncHost({
  useBuiltinGrammars: false,
  ignoreCase: false,
  resolveFile: file => file,
  readFileSync: file =>
    let idx = parseInt(file);
    if (idx.toString() !== file || idx < 0 || idx >= grammarParts.length) {
      throw new Error('tried to read non-existent ' + file);
    }
    return grammarParts[idx];
  }
});
```

For each callback hook, the behavior is:
- If the callback hook was provided execute the hook. The fallback behavior for the hook is provided via a `fallback` parameter on each hook.
- Otherwise, if a fallback host was provided, execute the method on the fallback hook.
- Otherwise, either do nothing or throw an error (depending on the hook).

**Comment by rbuckton** [comment]
I'll clean up the public API at some point in the future when I decide to publish a 3.x release.

File: src/host.ts [hunk 1]
```diff
@@ -41,13 +125,13 @@ export interface HostBaseOptions {
...
 
-    private useBuiltinGrammars: boolean;
-    private innerParser: Parser | undefined;
-    private knownGrammars: Map<string, string> | undefined;
+    private _innerParser: Parser | undefined;
+    private _knownGrammars: Map<string, string> | undefined;
+    private _useBuiltinGrammars: boolean;
 
-    constructor({ ignoreCase = ignoreCaseFallback, knownGrammars, useBuiltinGrammars = true }: HostBaseOptions = {}) {
+    constructor({ ignoreCase = false, knownGrammars, useBuiltinGrammars = true }: HostBaseOptions = {}) {
```
Comment (bakkot) [inline src/host.ts:132]: Built-in grammars don't make all that much sense for non-Node hosts, at least as currently implemented. Does it need to be part of the base API?

File: src/host.ts [hunk 2]
```diff
@@ -41,13 +125,13 @@ export interface HostBaseOptions {
 export abstract class HostBase {
     public readonly ignoreCase: boolean;
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/host.ts:126]: I believe I am depending on this in work-in-progress changes in ecmarkup-vscode. Case sensitivity affects not only internal representations (i.e. case folding for key comparisons), but also sort order in the UI, so having only `normalizeFile` isn't enough information.

File: src/host.ts [hunk 3]
```diff
@@ -57,46 +141,26 @@ export abstract class HostBase {
     }
 
     protected get parser(): Parser {
-        return this.innerParser
-            || (this.innerParser = this.createParser());
+        return this._innerParser
+            || (this._innerParser = this.createParser());
     }
 
     public normalizeFile(file: string) {
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/host.ts:148]: `normalizeFile` isn't used by `resolveFile` at all. It's used for generating case-insensitive keys for maps. Node does not have an API for getting the actual path for a given path on a case-insensitive file system. It's not a foolproof solution as its possible to mix and match case-sensitive and case-insensitive mount points in a single file system, but that's generally uncommon.

File: src/host.ts [hunk 4]
```diff
@@ -115,74 +179,102 @@ export abstract class HostBase {
...
-        if (!readFile) throw new Error("Operation cannot be completed synchronously");
-
         performance.mark("ioRead");
-        file = getLocalPath(file);
-        if (isUri(file)) return undefined; // TODO: support uris?
-        const result = readFile(file, wrapCancelToken(toCancelToken(cancelable)));
-        performance.measure("ioRead", "ioRead");
-        return result;
+        try {
+            file = getLocalPath(file);
```
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/host.ts:256]: I'll consider that. URI and disk path resolution works differently, so I mostly wrote it this way to abstract away the I/O operation for resolution.

File: src/host.ts (hunk 1 above)
Comment (rbuckton) [reply src/host.ts:132]: Built-in grammars are part of the package and have no dependency on Node being the host. That said, the built-in grammars are fairly old at this point. I'm using a slightly different mechanism in ecmarkup-vscode for writing derived grammars (essentially fetching the grammar from the ecma262 repo). I haven't decided what to do with this yet and will probably leave it in until I plan out what 3.x should look like.

File: src/host.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (rbuckton) [reply src/host.ts:256]: The issue is that NodeJS doesn't understand `file://` URLs unless they are in a parsed URL object, so I translate to a local path here. I can move that to the fallback read/write though.

[the remark this answers is not in the document]
Comment (bakkot) [reply src/host.ts]: Ah, neat, thanks.

File: src/host.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (bakkot) [reply src/host.ts:148]: > `normalizeFile` isn't used by `resolveFile`

Sorry, I meant this the other way around. That is: why do consumers of `normalizeFile` not use `resolveFile` instead? For map keys, it seems like you'd want to treat e.g. `a/./b` and `a/b` on the same (on Unix machines), which means doing resolution, not just case folding.

(On case-insensitive systems you couldn't use Node's `resolve` directly, but you could just have a `CaseInsensitiveNodeHost` or whatever that had a version of `resolveFile` which did case folding after Node's resolution.)

File: src/host.ts (hunk 1 above)
Comment (bakkot) [reply src/host.ts:132]: > Built-in grammars are part of the package and have no dependency on Node being the host.

Well, no, but their implementation kind of does: passing `useBuiltinGrammars = true` will [make](https://github.com/rbuckton/grammarkdown/blob/5a86b9fb55d00e4ed2a87d37d55b7723dc428b67/src/host.ts#L560) `resolveFile('ES6')` return the path on disk to `es2015.grammar`, but that path is only meaningful on Node hosts.

But if you're planning on revisiting it later anyway it makes sense to ignore it for now.

Comment (rbuckton) [reply src/host.ts:132]: It will return the result of `require.resolve`, which (IIRC) bundlers like WebPack overwrite.

Comment (bakkot) [reply src/host.ts:132]: Sure, it's just that a host which is not going to do some equivalent of `fs.readFile` can't really do much useful with those paths.

**Comment by rbuckton** [comment]
@bakkot: FYI, I've been working on the API documentation as well. It's not 100% complete yet, but many of the common API scenarios are documented at https://rbuckton.github.io/grammarkdown/

**Pull request merged**